File: 2022/day05.hs
-- Advent of Code 2022, day 5
--  Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let (sts,mvs) = parseAll pInput input
  putStrLn ("Part 1: " ++ part1 sts mvs)
  putStrLn ("Part 2: " ++ part2 sts mvs)

-- Parsing the input

type Stack = [Char]
type Stacks = M.Map Int Stack
type Move = (Int,Int,Int)

inputStacks :: String -> [String]
inputStacks s = let (s1,s2) = break (=='1') s
                in init (lines s1)

compressLine :: String -> String
compressLine s | length s <3 = []
               | otherwise   = (s!!1) : compressLine (drop 4 s)


stackMatrix :: [String] -> [Stack]
stackMatrix = map (dropWhile (==' ')) . transpose . map compressLine

stackMap :: String -> Stacks
stackMap = listMap 1 . stackMatrix . inputStacks

pMove :: Parser Move
pMove = do symbol "move"
           n <- natural
           symbol "from"
           fr <- natural
           symbol "to"
           to <- natural
           return (n,fr,to)

pInput :: Parser (Stacks,[Move])
pInput = do sts <- chunk
            line
            moves <- some pMove
            return (stackMap sts,moves)

-- Part 1

move :: Stacks -> Move -> Stacks
move st (k,n,m) =
  let xs = take k (st M.! n)
  in M.adjust (drop k) n (M.adjust (reverse xs ++) m st)

moves :: Stacks -> [Move] -> Stacks
moves = foldl move

stackTops :: Stacks -> String
stackTops = map head . M.elems

part1 :: Stacks -> [Move] -> String
part1 sts mvs = stackTops $ moves sts mvs

-- Part 2

move2 :: Stacks -> Move -> Stacks
move2 st (k,n,m) =
  let xs = take k (st M.! n)
  in M.adjust (drop k) n (M.adjust (xs ++) m st)

moves2 :: Stacks -> [Move] -> Stacks
moves2 = foldl move2

part2 :: Stacks -> [Move] -> String
part2 sts mvs = stackTops $ moves2 sts mvs

File: 2022/python/day05.py
# Advent of Code 2022, day 5
#   Venanzio Capretta

print("Advent of Code 2022, Day 5")

f = open("../input05")
input = f.readlines()
f.close()

# Parsing the input

# stacks 

def numberline(s):
  return (s[1]=='1')

stacks = [[] for i in range(9)]
while not numberline(input[0]):
  i = 0
  for k in range(0,len(input[0]),4):
    if input[0][k+1] != ' ' :
      stacks[i].append(input[0][k+1])
    i += 1
  input.pop(0)

for s in stacks:
  s.reverse()

# moves

# eliminates numbers and blanks
input.pop(0)
input.pop(0)

def move(s):
  (x,s1) = s.split("move ",1)
  (num,s2) = s1.split(" from ",1)
  (frs,tos) = s2.split(" to ",1)
  return (int(num),int(frs),int(tos))

moves = []
for m in input:
  moves.append(move(m))
 
# Part 1

stacks1 = [l.copy() for l in stacks]

for (n,s,t) in moves:
  for i in range(n):
    stacks1[t-1].append(stacks1[s-1].pop())

def answer(sts):
  ans = ""
  for l in sts:
    ans += l[-1]
  return ans

print("Part 1: ")
print(answer(stacks1))

# Part 2

stacks2 = [l.copy() for l in stacks]

for (n,s,t) in moves:
  l = stacks2[s-1][-n:]
  stacks2[s-1] = stacks2[s-1][:-n]
  stacks2[t-1] += l

print("Part 2: ")
print(answer(stacks2))

